runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/getdata/ThrottlingGetDataMetricTracker.java

```diff
@@ -0,0 +1,146 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.dataflow.worker.windmill.client.getdata;
+
+import com.google.auto.value.AutoValue;
+import java.io.PrintWriter;
+import java.util.concurrent.atomic.AtomicInteger;
+import javax.annotation.concurrent.ThreadSafe;
+import org.apache.beam.runners.dataflow.worker.util.MemoryMonitor;
+import org.apache.beam.sdk.annotations.Internal;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.annotations.VisibleForTesting;
+
+/**
+ * Wraps GetData calls that tracks metrics for the number of in-flight requests and throttles
+ * requests when memory pressure is high.
+ */
+@Internal
+@ThreadSafe
+public final class ThrottlingGetDataMetricTracker {
+  private final MemoryMonitor gcThrashingMonitor;
+  private final GetDataMetrics getDataMetrics;
+
+  public ThrottlingGetDataMetricTracker(MemoryMonitor gcThrashingMonitor) {
+    this.gcThrashingMonitor = gcThrashingMonitor;
+    this.getDataMetrics = GetDataMetrics.create();
+  }
+
+  /**
+   * Tracks a GetData call. If there is memory pressure, may throttle requests. Returns an {@link
+   * AutoCloseable} that will decrement the metric after the call is finished.
+   */
+  public AutoCloseable trackSingleCallWithThrottling(Type callType) {
+    gcThrashingMonitor.waitForResources(callType.debugName);
+    AtomicInteger getDataMetricTracker = getDataMetrics.getMetricFor(callType);
+    getDataMetricTracker.getAndIncrement();
+    return getDataMetricTracker::getAndDecrement;
+  }
+
+  /**
+   * Tracks heartbeat request metrics. Returns an {@link AutoCloseable} that will decrement the
+   * metric after the call is finished.
+   */
+  public AutoCloseable trackHeartbeats(int numHeartbeats) {
+    getDataMetrics
+        .activeHeartbeats()
+        .getAndUpdate(currentActiveHeartbeats -> currentActiveHeartbeats + numHeartbeats);
+    return () ->
+        getDataMetrics.activeHeartbeats().getAndUpdate(existing -> existing - numHeartbeats);
+  }
+
+  public void printHtml(PrintWriter writer) {
+    writer.println("Active Fetches:");
+    getDataMetrics.printMetrics(writer);
+  }
+
+  @VisibleForTesting
+  GetDataMetrics.ReadOnlySnapshot getMetricsSnapshot() {
+    return getDataMetrics.snapshot();
+  }
+
+  public enum Type {
+    STATE("GetStateData"),
+    SIDE_INPUT("GetSideInputData"),
+    HEARTBEAT("RefreshActiveWork");
+    private final String debugName;
+
+    Type(String debugName) {
+      this.debugName = debugName;
+    }
+
+    public final String debugName() {
+      return debugName;
+    }
+  }
+
+  @AutoValue
+  abstract static class GetDataMetrics {
```

Comment:
I don't think this adds much since nothing is ever set differently and ThrottlingGetDataMetricTracker is already a pretty focused class. It would be less code to just inline the AtomicIntegers as members in the tracker

Reply:
removed